hasher-matcher-actioner/hmalib/common/timebucketizer.py

```diff
@@ -0,0 +1,161 @@
+        raise NotImplementedError
+
+    @classmethod
+    def from_csv(cls: t.Type[Self], value: t.List[str]) -> Self:
+        raise NotImplementedError
+
+
+T = t.TypeVar("T", bound=CSViable)
+
+
+class TimeBucketizer(t.Generic[T]):
+
+    # TODO Make this generic, so we can have TimeBucket[Tuple[int, int]] or TimeBucket[MoreComplexObject].
+    def __init__(
+        self, bucket_width: datetime.timedelta, storage_path: str, type: str, id: str, buffer_size: int = MAX_BUFFER_SIZE
+    ):
+        """
+        Divides the day into 24h / bucket_width buckets. When add_record
+        is called, based on current time, appends record to current bucket.
+        Each bucket's data is stored as a serialized CSV file.
+        Note that there may be multiple instances of TimeBucket writing to the
+        same file system so add some uniqueness for this instance.
+
+        Say your storage path is /var/data/threatexchange/timebuckets
+        The file for current datetime (2022/02/08/20:49 UTC) for a bucket_width
+        of 5 minutes should be "/var/data/threatexchange/timebuckets/<type>/2022/02/08/20:45/<unique_id>.csv"
+
+        The first <type> allows you to use the same storage folder for
+        say hashes and matches.
+        The second <unique_id> allows you to have multiple instances running
+        (eg. multiple lambdas) and writing to the same file system.
+        """
+
+        if bucket_width.total_seconds() < 60 or bucket_width.total_seconds() % 60 != 0:
+            raise Exception("Please ensure timedelta is atleast a minute long.")
+
+        if SECONDS_PER_DAY % bucket_width.total_seconds() != 0:
+            raise Exception(
+                "Time Delta must be equally divisible into buckets based on a 24 hour clock"
+            )
+
+        self.bucket_width = bucket_width
+        self.start, self.end = self._calculate_bucket_endpoints(datetime.datetime.now(), bucket_width)
+        self.storage_path = storage_path
+        self.id = id
+        self.type = type
+
+        self.buffer_size = buffer_size
+        self.buffer: t.List[T] = []
+
+
+    def _calculate_bucket_endpoints(self, time: datetime.datetime, bucket_width: datetime.timedelta):
+        now = time
+        rounded = now - (now - datetime.datetime.min) % bucket_width
+        return (rounded, rounded + self.bucket_width)
+
+    def add_record(self, record: T) -> None:
+        """
+        Adds the record to the current bucket.
+        """
+        if len(self.buffer) >= MAX_BUFFER_SIZE or self.end <= datetime.datetime.now():
+            self._flush()
+        self.buffer.append(record)
+
+    def destroy_remaining(self) -> None:
+        """
+        Used to trigger a force flush of remaining data stored inside buffer
+        """
+        if not len(self.buffer):
+            return
+        self._flush()
+
+    # Change back to t.List[T] after other method is implemented
+    def get_records(self) -> t.List[str]:
```

Comment:
This "was" only used for testing. Is not anymore. Can you remove this method?

It is supplanted by get_file_contents.